### PR TITLE
Move to use metadata section in extensions manifest

### PR DESCRIPTION
**Pull request by maxandersen**
Why:

 * to have an openended set of data we can use in aggregation
   and UI's.
 * labels notion caused too much ambiguity; move to more explicit keywords.

This change addreses the need by:

 * add metadata to Extension.java
 * have extension mojo "migrate" all existing extensions
 * move labels to use keywords

note: existing code still uses keywords. Follow fix will be done to
remove the old extensions.json together with notion of labels.

### CODE REVIEW

File: independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java [hunk 1]
```diff
@@ -153,11 +158,41 @@ public void execute() throws MojoExecutionException {
...
+        
         try (BufferedWriter bw = Files.newBufferedWriter(output.resolve(BootstrapConstants.EXTENSION_PROPS_JSON_FILE_NAME))) {
             extObject.writeTo(bw, WriterConfig.PRETTY_PRINT);
         } catch (IOException e) {
             throw new MojoExecutionException("Failed to persist " + output.resolve(BootstrapConstants.EXTENSION_PROPS_JSON_FILE_NAME), e);
         }
     }
+
+	private void transformLegacyToNew(final Path output, JsonObject extObject) throws MojoExecutionException {
+		JsonObject metadata = new JsonObject();
```
Comment (gastaldi) [inline independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java]: No need to create a new here, better do that in an `else` below

File: independent-projects/tools/platform-descriptor-api/src/test/java/io/quarkus/dependencies/test/ExtensionSerializerTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+		
+		String[] input = new String[] { "what", "not", "to", "do"};
+		
+		e.setLabels(input);
+		
+		List<String> list = (List<String>) e.getMetadata().get("keywords");
+		
+		assertNotNull(list);
+		
+		assertArrayEquals(input, list.toArray(new String[0]));
```
Comment (gastaldi) [inline independent-projects/tools/platform-descriptor-api/src/test/java/io/quarkus/dependencies/test/ExtensionSerializerTest.java]: You can use Assertj instead `assertThat(list).containsExactly(input)`

Comment (gastaldi) [reply independent-projects/tools/platform-descriptor-api/src/test/java/io/quarkus/dependencies/test/ExtensionSerializerTest.java]: Also why an additional line break per statement?

**Review by gastaldi — APPROVED** [review]
LGTM, added some minor comments

File: independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java [hunk 3]
```diff
@@ -153,11 +158,41 @@ public void execute() throws MojoExecutionException {
...
+        // Only useful if wanting to write back to original location
+        /**  
+        Path source = output.resolve("../../../src/main/resources/META-INF/" + BootstrapConstants.EXTENSION_PROPS_JSON_FILE_NAME);
+        if (source.toFile().exists()) {
+        	try (BufferedWriter bw = Files.newBufferedWriter(source)) {
+                extObject.writeTo(bw, WriterConfig.PRETTY_PRINT);
+            } catch (IOException e) {
+                throw new MojoExecutionException("Failed to persist " + source, e);
+            }	
+        }*/
```
Comment (aloubyansky) [inline independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java]: This should go away.

File: independent-projects/tools/platform-descriptor-api/src/main/java/io/quarkus/dependencies/Extension.java [hunk 4]
```diff
@@ -98,11 +102,15 @@ public Extension setClassifier(String classifier) {
     }
 
     public String[] getLabels() {
-        return labels;
+    	List<String> keywords = (List<String>) getMetadata().get("keywords");
+    	if(keywords != null) {
+    		return (String[]) keywords.toArray(new String[keywords.size()]);
```
Comment (aloubyansky) [inline independent-projects/tools/platform-descriptor-api/src/main/java/io/quarkus/dependencies/Extension.java]: This looks like not been formatted.

File: independent-projects/tools/platform-descriptor-api/src/test/java/io/quarkus/dependencies/test/ExtensionSerializerTest.java (hunk 2 above)
Comment (maxandersen) [reply independent-projects/tools/platform-descriptor-api/src/test/java/io/quarkus/dependencies/test/ExtensionSerializerTest.java]: I don't have/need assertj anywhere else and this code using labels will die in next update so didn't bother :)

File: independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java (hunk 1 above)
Comment (maxandersen) [reply independent-projects/bootstrap/maven-plugin/src/main/java/io/quarkus/maven/ExtensionDescriptorMojo.java]: fixed.

File: independent-projects/tools/platform-descriptor-api/src/test/java/io/quarkus/dependencies/test/ExtensionSerializerTest.java (hunk 2 above)
Comment (maxandersen) [reply independent-projects/tools/platform-descriptor-api/src/test/java/io/quarkus/dependencies/test/ExtensionSerializerTest.java]: btw. not sure why the additional line breaks - fixed it.

**Review by aloubyansky — APPROVED** [review]

**Pull request merged**